cmd/gateway/s3x/gateway-s3x-ledger.go

```diff
@@ -15,418 +17,84 @@ Any claiming or releasing of locks should be done in the public setter+getter fu
-		IpfsHash: hash,
-	}
-	return le.putLedger(ledger)
-}
-
-// UpdateBucketHash is used to update the ledger bucket entry
-// with a new IPFS hash
-func (le *LedgerStore) UpdateBucketHash(name, hash string) error {
-	le.Lock()
-	defer le.Unlock()
-	ledger, err := le.getLedger()
-	if err != nil {
-		return err
-	}
-	if !le.bucketExists(ledger, name) {
-		return ErrLedgerBucketDoesNotExist
-	}
-	entry := ledger.Buckets[name]
-	entry.IpfsHash = hash
-	ledger.Buckets[name] = entry
-	return le.putLedger(ledger)
-}
-
-// RemoveObject is used to remove a ledger object entry from a ledger bucket entry
-func (le *LedgerStore) RemoveObject(bucketName, objectName string) error {
-	le.Lock()
-	defer le.Unlock()
-	ledger, err := le.getLedger()
-	if err != nil {
-		return err
-	}
-	if err := le.objectExists(ledger, bucketName, objectName); err != nil {
-		return err
-	}
-	delete(ledger.Buckets[bucketName].Objects, objectName)
-	return nil
-}
-
-// AddObjectToBucket is used to update a ledger bucket entry with a new ledger object entry
-func (le *LedgerStore) AddObjectToBucket(bucketName, objectName, objectHash string) error {
-	le.Lock()
-	defer le.Unlock()
-	ledger, err := le.getLedger()
-	if err != nil {
-		return err
-	}
-	if !le.bucketExists(ledger, bucketName) {
-		return ErrLedgerBucketDoesNotExist
-	}
-	// prevent nil map panic
-	if ledger.GetBuckets()[bucketName].Objects == nil {
-		bucket := ledger.Buckets[bucketName]
-		bucket.Objects = make(map[string]LedgerObjectEntry)
-		ledger.Buckets[bucketName] = bucket
-	}
-	ledger.Buckets[bucketName].Objects[objectName] = LedgerObjectEntry{
-		Name:     objectName,
-		IpfsHash: objectHash,
-	}
-	return le.putLedger(ledger)
-}
-
-// DeleteBucket is used to remove a ledger bucket entry
-func (le *LedgerStore) DeleteBucket(name string) error {
-	le.Lock()
-	defer le.Unlock()
-	ledger, err := le.getLedger()
-	if err != nil {
-		return err
-	}
-	if ledger.GetBuckets()[name].Name == "" {
-		return ErrLedgerBucketDoesNotExist
-	}
-	delete(ledger.Buckets, name)
-	return le.putLedger(ledger)
-}
-
 // Close shuts down the ledger datastore
-func (le *LedgerStore) Close() error {
-	le.Lock()
-	defer le.Unlock()
-	return le.ds.Close()
+func (ls *ledgerStore) Close() error {
+	//todo: clean up caches
+	return ls.ds.Close()
 }
 
 /////////////////////
 // GETTER FUNCTINS //
 /////////////////////
 
-// GetObjectParts is used to return multipart upload parts
-func (le *LedgerStore) GetObjectParts(id string) ([]ObjectPartInfo, error) {
-	le.RLock()
-	defer le.RUnlock()
-	ledger, err := le.getLedger()
+// GetObjectInfos returns a list of ordered ObjectInfos with given prefix ordered by name
+func (ls *ledgerStore) GetObjectInfos(ctx context.Context, bucket, prefix, startsFrom string, max int) ([]ObjectInfo, error) {
+	defer ls.locker.read(bucket)()
+	b, err := ls.getBucketLoaded(ctx, bucket)
 	if err != nil {
 		return nil, err
 	}
-	if err := le.multipartExists(ledger, id); err != nil {
-		return nil, err
+	var names []string
```

Comment:
we should initialize this to `max` that way on large queries we have much better memory usage. Probably do something like:

`var names = make([]string, 0, max)`

Reply:
But it will take more memory for short queries.

Reply:
good point